kafka/signed_token_redeem_handler.go

```diff
@@ -151,25 +214,58 @@ func SignedTokenRedeemHandler(
 			redeemedTokenResults = append(redeemedTokenResults, avroSchema.RedeemResult{
 				Issuer_name:     "",
 				Issuer_cohort:   0,
-				Status:          redeemUnverified,
+				Status:          avroSchema.RedeemResultStatusUnverified,
 				Associated_data: request.Associated_data,
 			})
 			continue
 		} else {
-			logger.Trace().Msgf("request %s: validated", tokenRedeemRequestSet.Request_id)
+			logger.Info().Msg(fmt.Sprintf("request %s: validated", tokenRedeemRequestSet.Request_id))
 		}
-		if err := server.RedeemToken(verifiedIssuer, &tokenPreimage, request.Binding); err != nil {
-			logger.Error().Err(fmt.Errorf("request %s: token redemption failed: %w",
-				tokenRedeemRequestSet.Request_id, err)).
-				Msg("signed token redeem handler")
+		redemption, equivalence, err := server.CheckRedeemedTokenEquivalence(verifiedIssuer, &tokenPreimage, string(request.Binding), msg.Offset)
+		if err != nil {
+			message := fmt.Sprintf("request %s: failed to check redemption equivalence", tokenRedeemRequestSet.Request_id)
+			processingResult, errorResult := avroRedeemErrorResultFromError(
+				message,
+				msg,
+				producer,
+				tokenRedeemRequestSet.Request_id,
+				int32(avroSchema.RedeemResultStatusError),
+				log,
+			)
+			MayEmitIfPermanent(processingResult, errorResult, producer, log)
+			return NilIfPermanent(errorResult)
+		}
+
+		// Continue if there is a duplicate
+		switch equivalence {
+		case cbpServer.IDEquivalence:
+			redeemedTokenResults = append(redeemedTokenResults, avroSchema.RedeemResult{
+				Issuer_name:     "",
+				Issuer_cohort:   0,
+				Status:          avroSchema.RedeemResultStatusDuplicate_redemption,
+				Associated_data: request.Associated_data,
+			})
+			continue
+		case cbpServer.BindingEquivalence:
+			redeemedTokenResults = append(redeemedTokenResults, avroSchema.RedeemResult{
+				Issuer_name:     "",
+				Issuer_cohort:   0,
+				Status:          avroSchema.RedeemResultStatusIdempotent_redemption,
+				Associated_data: request.Associated_data,
+			})
+			continue
+		}
+
+		if err := server.PersistRedemption(*redemption); err != nil {
+			logger.Error().Err(err).Msgf("request %s: token redemption failed: %e", tokenRedeemRequestSet.Request_id, err)
 			if strings.Contains(err.Error(), "Duplicate") {
 				logger.Error().Err(fmt.Errorf("request %s: duplicate redemption: %w",
 					tokenRedeemRequestSet.Request_id, err)).
 					Msg("signed token redeem handler")
 				redeemedTokenResults = append(redeemedTokenResults, avroSchema.RedeemResult{
 					Issuer_name:     "",
 					Issuer_cohort:   0,
-					Status:          redeemDuplicateRedemption,
+					Status:          avroSchema.RedeemResultStatusDuplicate_redemption,
```

Comment:
is this always guaranteed to be a duplicate vs idempotent? in what case do we get here, there was a race and some other process stored a token redemption with the same key?

Reply:
Definitely possible for there to be a race condition. I'll account for it.